core/controllers/reader.py

```diff
@@ -1792,20 +2223,90 @@ def put(self, exploration_id):
         self.render_json(self.values)
 
 
-class LearnerAnswerDetailsSubmissionHandler(base.BaseHandler):
+class LearnerAnswerDetailsSubmissionHandlerNormalizedPayloadDict(TypedDict):
+    """Dict representation of LearnerAnswerDetailsSubmissionHandler's
+    normalized_payload dictionary.
+    """
+
+    state_name: Optional[str]
+    interaction_id: str
+    answer: Union[str, int, Dict[str, str], List[str]]
+    answer_details: str
+
+
+class LearnerAnswerDetailsSubmissionHandler(
+    base.BaseHandler[
+        LearnerAnswerDetailsSubmissionHandlerNormalizedPayloadDict,
+        Dict[str, str]
+    ]
+):
     """Handles the learner answer details submission."""
 
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+    URL_PATH_ARGS_SCHEMAS = {
+        'entity_type': {
+            'schema': {
+                'type': 'basestring',
+                'choices': [
+                    feconf.ENTITY_TYPE_EXPLORATION,
+                    feconf.ENTITY_TYPE_QUESTION
+                ]
+            }
+        },
+        'entity_id': {
+            'schema': {
+                'type': 'basestring',
+                'validators': [{
+                    'id': 'is_regex_matched',
+                    'regex_pattern': constants.ENTITY_ID_REGEX
+                }]
+            }
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'PUT': {
+            'state_name': {
+                'schema': {
+                    'type': 'basestring'
+                },
+                'default_value': None
```

Comment:
Why can this be `None`?

Reply:
Because if `entity_type == feconf.ENTITY_TYPE_EXPLORATION` condition is not true then we don't need state_name.

Reply:
Hmm, that makes sense, the ideal solution would probably be to split up the handler into two, one for exp and one for question.